File: Hash Table/hashtable.cpp
#include <iostream>
#include <vector>
using namespace std;
#define size 10
// Collision resolution by chaining.


class hashTable{

    private:
        vector<int> table_[size];

        int h(int n){
            int index = n % size;
            return index;
        }
    
    public:

        void insert(int arr[], int n){
            for(int x=0; x < n; x++){
                insert(arr[x]);
                cout<<arr[x];
            }
        }

        void insert(int val){

            int index = h(val);

            table_[index].push_back(val);
        }

        bool search(int val){

            for ( int i = 0; i < size; i++ ){
                vector<int>::iterator itr;
                int counter = 0;
                for(itr=table_[i].begin(); itr < table_[i].end(); itr++){
                    
                    if(*itr == val){
                        cout<<*itr<<" is Found at index"<<i<<"x"<<counter<<endl;
                        return true;
                    }
                    counter++;
                }
            }

            return false;

        }

};



int main(){

    hashTable t;
    t.insert(11);
    t.insert(1);
    t.insert(10);
    t.search(1);
    t.search(10);

    return 0;
}

File: Hash Table/hashtable.py
class HashTable:
    
    def __init__(self, size):
        
        self.table = []
        for x in range (0, size):
            self.table.append([])

        if size < 0:
            print("Size cannot be negative")

        else:
            self.size = size

    def h(self, val):

        index = val % self.size
        return index
    
    def insertMany(self, arr):

        for x in arr:
            self.insert(x)

    def insert(self, val):

        i = self.h(val)
        self.table[i].append(val)

    def search(self, val):

        i = self.h(val)
        counter = 0

        for x in self.table[i]:
            if( x == val ):
                return f"Found in {i}x{counter}"
            
            counter+=1
        
        return "Not Found"
    

hash = HashTable(10)
hash.insertMany([1,3,4,1,5,7, 77, 6,43,12,6,85,12,5,62,2,10,0])
print(hash.table)
